### PR TITLE
Support exporting environment-specific metrics

### PR DESCRIPTION
**Pull request by mdawn65**
Issue: #139 

### Summary
This PR adds support for environment-specific metrics collection and logging in SkyRL-Gym environments. 

- Added `metrics` of type `Dict[str,Any]` to `BaseTextEnvStepOutput`
- Added `env_metrics` of type `Optional[List[Dict[str, Any]]]` to `GeneratorOutput`
- Logic added in `concatenate_generator_outputs` in `utils.py`
- Logic added in `generate()` and `generate_batched()` to collect metrics from environment steps
- `eval()` computes aggregation by **mean** across evaluation batches
- Aggregated metrics are logged to wandb under the `env/*` name
- Example environment: `GSM8k`
   - Added `_calculate_metrics()` method for the metrics: response_length, word_count, answer_accuracy
- Added 4 unit tests (non-batched and batched, compatibility with envs that don't report metrics, concatenation logic, edge cases)

### CODE REVIEW

File: skyrl-train/skyrl_train/generators/skyrl_gym_generator.py [hunk 1]
```diff
@@ -311,6 +316,7 @@ async def generate(self, input_batch: GeneratorInput) -> GeneratorOutput:
         stop_reasons = [output[2] for output in all_outputs]
         loss_masks = [output[3] for output in all_outputs]
         prompt_token_ids = [output[4] for output in all_outputs]
+        env_metrics = [output[6] for output in all_outputs]
```
[the remark this answers is not in the document]
Comment (tyler-griggs) [reply skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:380]: Don't worry about this, it is something I will handle separately.

File: skyrl-train/skyrl_train/trainer.py [hunk 2]
```diff
@@ -177,14 +180,38 @@ async def eval(self) -> Dict[str, float]:
 
         # 3. Calculate overall metrics across all datasets
         overall_avg_score, overall_pass_at_n = get_metrics_from_generator_output(concat_generator_outputs, concat_uids)
+
+        # 4. Aggregate environment metrics for logging
```
Comment (tyler-griggs) [inline skyrl-train/skyrl_train/trainer.py]: We won't want to handle environment metric aggregation in the training stack because the metrics will be pretty environment specific, and it might surprise env develops when the training stack quietly aggregates its metrics. It should be the responsibility of the environment to product Dict[str, Any] object that can be directly exported by wandb (and other loggers) so that the training stack just directly exports it (see the rollout_metrics as an example). Let me know if I'm misunderstanding something here though!

File: skyrl-train/skyrl_train/generators/skyrl_gym_generator.py [hunk 3]
```diff
@@ -341,11 +347,12 @@ async def generate(self, input_batch: GeneratorInput) -> GeneratorOutput:
             "stop_reasons": stop_reasons,
             "rollout_metrics": rollout_metrics,
             "rollout_logprobs": rollout_logprobs,
+            "env_metrics": env_metrics,
```
Comment (tyler-griggs) [inline skyrl-train/skyrl_train/generators/skyrl_gym_generator.py]: Can we actually bundle the env_metrics into `rollout_metrics` and only return `rollout_metrics` in `GeneratorOutput`. Long-term, I believe we will actually change this name to just `metrics` or `generation_metrics` to be more generic, but I'd prefer to keep the `GeneratorOutput` interface unchanged.

Comment (tyler-griggs) [reply skyrl-train/skyrl_train/generators/skyrl_gym_generator.py]: Relatedly, we should probably prefix all the env_metrics keys from the environment with the string "environment/" so that it's well-organized in wandb (or other logging tools)

**Review by tyler-griggs — COMMENTED** [review]
This is looking great! I left a couple comments. Also, it would be great to launch a simple training run (examples/gsm8k/run_gsm8k.sh with a 0.5B model) and upload metrics to wandb (you can create a free personal account and get wandb api key to use), then attach a screenshot of the reported example metrics (`response_length`, `word_count`, etc.) just to verify that everything looks right

File: skyrl-train/skyrl_train/trainer.py (hunk 2 above)
Comment (mdawn65) [reply skyrl-train/skyrl_train/trainer.py]: Hey Tyler, after we aggregate all of the metrics from generation, the output of `concat_all_env_metrics` is `List[Dict[str, Any]]`. Should I add an aggregate function in env.py (for example, to `gsm8k/env.py`), or should I leave the data as is?

File: skyrl-train/skyrl_train/generators/skyrl_gym_generator.py (hunk 3 above)
Comment (mdawn65) [reply skyrl-train/skyrl_train/generators/skyrl_gym_generator.py]: Hi Tyler, I've addressed the comments on this new PR: https://github.com/NovaSky-AI/SkyRL/pull/189.
